docs/adr/adr-11-blocksync-overhaul.md

```diff
@@ -0,0 +1,329 @@
+simple hash to data requests. The top-level index maps any hash to any block CARv1file so that FNs/BNs can quickly
+serve requests.
+
+- __FNs/BNs run a single instance of [`DAGStore`](https://github.com/filecoin-project/dagstore) to manage CARv1 block files.__
+DAGStore provides both the top-level indexing and CARv2 based indexing per each CARv1 file. In essence, it's an engine
+for managing any CAR files with indexing, convenient abstractions, tools, recovery mechanisms, etc.
+
+- __LNs DASing remains untouched__. Both the networking protocol and storage for LNs remains untouched as it fulfills
+    the requirements. This includes Bitswap as backbone protocol for requesting samples and global Badger KVStore.
+
+- __New libp2p based Exchange protocol is introduced for data/share exchange purposes.__ FNs/BNs servers LNs clients.
+
+### Detailed Design
+
+> All the comments on the API definitions should be preserved and potentially improved by implementations.
+
+#### Block/EDS Storage
+
+The new block storage design is solely additive. Meaning that all the existing storage related components and functionality
+are kept with additional components introduced. Altogether, existing and new components will be recomposed to serve the
+foundation of our improved block storage subsystem.
+
+The central data structure representing Celestia block data is EDS(`rsmt2d.ExtendedDataSquare`) and the new storage design
+is focused around storing entire EDSes as a whole rather than a set of individual chunks s.t. storage subsystem
+can handle storing and streaming/serving blocks of 4mb sizes and more.
+
+##### EDS Serde
+
+Storing EDS as a whole requires EDS (de)serialization. For this the [CAR format](https://ipld.io/specs/transport/car) is
+chosen.
+
+###### `share.WriteEDS`
+
+To write EDS into a stream/file `WriteEDS` is introduced.
+// TODO Describe logic
+
+NOTE: CAR provides [a utility](https://github.com/ipld/go-car/blob/master/car.go#L47) to serialize any DAG into the file and
+there is a way to serialize EDS into DAG(`share/ipld.ImportShares`). This approach is the simplest and traverses shares
+and Merkle Proofs in depth-first manner packing them in a CAR file. However, this is incompatible with the requirement
+to truncate the CAR file to get strictly the first quadrant out of it without NMT proofs, so serialization must be
+different from the utility to support that.
+
+NOTE2: Alternatively to `WriteEDS`, and `EDSReader` could be introduced to make EDS-to-stream handling more idiomatic
+and efficient in some cases, with the cost of more complex implementation.
+
+```go
+// WriteEDS writes whole EDS into given io.Writer as CARv1 file.
+// All its shares and recomputed NMT proofs.
+func WriteEDS(context.Context, *rsmt2d.ExtendedDataSquare, io.Writer) error
+```
+
+###### `share.ReadEDS`
+
+To read EDS out of stream/file `ReadEDS` is introduced. Internally, it
+
+- Imports EDS with an empty pre-allocated slice. NOTE: Size can be taken from DataRoot
```

Comment:
Is the dataroot not just a merkle hash? I know that if we have the length of rowroots/colroots we know how much data there is, but can you help me understand what we can do to go from root -> length here?

Reply:
Do you mean by first opening the CAR? (since the dataroot is in the top level index)

Reply:
Also by the way. I think we can't use dataroot as a primary index, since it is not unique (I thought it was, but it isn't, in the case of empty data, right?). I am currently using it as a filename but it isn't working for this reason

Reply:
It's not unique, but we don't need to store each empty block as its the same data.

Reply:
ok that was the secondary question, can you answer the one about how we get to the size?

Reply:
DataRoot equals to DAHeader in this case(The DataRoot comes from [share.Root](https://github.com/celestiaorg/celestia-node/blob/main/service/share/share.go#L35) and I should clarify that). And from DAHeader we can get the height.

Reply:
Clarified